### PR TITLE
Migrate Search overlays to regular layer

### PR DESCRIPTION
**Pull request by gberaudo**
Should be rebased and merged after #2590.

[Test](https://mf-geoadmin3.dev.bgdi.ch/gberaudo_search_layer/src/?lang=en&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&dev3d=true&catalogNodes=532&layers=ch.swisstopo.vec200-names-namedlocation&lon=7.01749&lat=46.51861&elevation=76984&heading=359.927&pitch=-89.998)

### CODE REVIEW

**Comment by gberaudo** [comment]
In this PR, a regular `ol.layer.Layer` is used instead of the `ol.Overlay`.
The layer is detected by OL3-Cesium and automatically synchronized to Cesium.

It looks good and I think we can merge it.

However, there is still substantial work to do for ensuring the layer stays on top of the other vector layers.

**Comment by gjn** [comment]
The `always on top` need to be done before merging into master. Put to [WIP] because of it.

**Comment by oterral** [comment]
The problem is we have unmanaged layer everywhere  (draw  interaction for example or geolocation). It was a nice feature to don't have to manage this `always on top` thing. 

Don't you think it will be better to have access of unmanaged layer from ol and use the `Vector Synchronizer` to display it on cesium ?

File: src/components/MarkerOverlayService.js [hunk 1]
```diff
@@ -6,10 +6,30 @@ goog.provide('ga_marker_overlay_service');
...
+
+      /** @const */
+      var style = new ol.style.Style({
+        image: new ol.style.Icon({
+          anchor: [0.5, 46],
+          anchorXUnits: 'fraction',
+          anchorYUnits: 'pixels',
+          src: 'img/marker.png'
+        })
+      });
```
Comment (oterral) [inline src/components/MarkerOverlayService.js]: Usually we define style in `gaStyleFactory` service. See https://github.com/geoadmin/mf-geoadmin3/blob/master/src/components/StylesService.js#L138

**Comment by gberaudo** [comment]
@oterral I opened an issue in OL3 so that we discuss what can be done:
https://github.com/openlayers/ol3/issues/4053

I propose several solutions; is the first one what you have in mind?

**Comment by gjn** [comment]
From those proposals, I tend to think that 2 or 3 is the most reasonable. It fits well with css z-index approach (and ol3 could even read the z-index for layers from css....)

1) does not cover the ordering of multiple unmanaged layers, which would need to be managed by the user (rare use case, but might exists). Also, unmanaged layers are managed again.
2) doesn't break current usage of unmanaged layers
3) probably ideal, but will break current usage of unmanaged layers (clients need to adapt code)

**Comment by oterral** [comment]
thx @gberaudo, nice summary.

Yes 1) seems the easisest way.

The problem of 2) and 3) is you have a Collection where the order doesn't reflect what you see and that can be a problem when you want to insert a Layer somewhere in the Collection or if you specify 2 layers with the same z-index or 2 layers without z-index. It seems to me that could be a mess to manage :).

Unmaniged layers are temporary layers, they are not displayed all the time. They are often display for a low period of time.

**Comment by cedricmoullet** [comment]
@gberaudo could you please rebase ?

**Comment by gberaudo** [comment]
The second proposal was merged in OL3.
Work was done on ol3-cesium to fix the blue flash when adding/updating/removing layers and using the Z-index available from OL3. The PR review https://github.com/openlayers/ol3-cesium/pull/241 is ongoing.
FlyTo 3D animation was added for locations and features.
The PR was rebased on master.

**Comment by oterral** [comment]
Very nice!!! can you create a test link pls `make deploybranch`

**Comment by gberaudo** [comment]
Here is a link: https://mf-geoadmin3.dev.bgdi.ch/gberaudo_search_layer/src/?lang=en&topic=ech&bgLayer=ch.swisstopo.pixelkarte-farbe&dev3d=true&catalogNodes=532&layers=ch.swisstopo.vec200-names-namedlocation&lon=7.01749&lat=46.51861&elevation=76984&heading=359.927&pitch=-89.998

**Comment by gberaudo** [comment]
This PR is now ready for review.

**Comment by cedricmoullet** [comment]
The markers are supposed to be on the terrain ?
I can see this issue:
image

**Comment by gberaudo** [comment]
Which marker is it? All locations I have tried so far were displayed
correctly on terrain.

**Comment by oterral** [comment]
@gberaudo I got one : search for `Alpiner Gipfel Stockhorn (VS) - Zermatt` then  if you move the map, the marker is displaying in the moutain.
stochhorn

**Comment by cedricmoullet** [comment]
"Rue des Berges 37 Payerne"

**Comment by gberaudo** [comment]
I was not able to reproduce the issue at Payerne and could not find the `Alpiner Gipfel Stockhorn (VS) - Zermatt` location.
However I could find a case, with `Gebaeude Einzelhaus Restaurant Alpenruh (BE) - Reichenbach im Kandertal`, where the symbol is cut by the mountain. For this case it is the expected behaviour as we want to have symbols behind a mountain hidden by it.

It might be the same cause in Payern if for some reason the depth was not updated correctly due to some glitch with the terrain or LOD race condition. @cedricmoullet are you able to consistently reproduce the issue?

payerne
through_mountain

**Comment by cedricmoullet** [comment]
No i can't reproduce....

2015-09-09 17:59 GMT+02:00 gberaudo <notifications@github.com>:

> I was not able to reproduce the issue at Payerne and could not find the Alpiner
> Gipfel Stockhorn (VS) - Zermatt location.
> However I could find a case, with Gebaeude Einzelhaus Restaurant Alpenruh
> (BE) - Reichenbach im Kandertal, where the symbol is cut by the mountain.
> For this case it is the expected behaviour as we want to have symbols
> behind a mountain hidden by it.
>
> It might be the same cause in Payern if for some reason the depth was not
> updated correctly due to some glitch with the terrain or LOD race
> condition. @cedricmoullet <https://github.com/cedricmoullet> are you able
> to consistently reproduce the issue?
>
> [image: payerne]
> <https://cloud.githubusercontent.com/assets/7294662/9767041/759bdadc-571c-11e5-98c4-f274334e1dfd.png>
> [image: through_mountain]
> <https://cloud.githubusercontent.com/assets/7294662/9767042/75a73e04-571c-11e5-9727-7925a409b0b3.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2604#issuecomment-138956352>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://t.sidekickopen16.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XYgf6M1WN1qfVVMRYFRqW8qlQyH56dByMf1q8vLz02?t=http%3A%2F%2Ftwitter.com%2Fcedricmoullet&si=6057645868318720&pi=5b97c8f5-2860-4da3-904c-ee753ebe4e09>

**Comment by cedricmoullet** [comment]
@gjn @oterral is is ready to merge ?

File: src/components/search/SearchTypesDirectives.js [hunk 2]
```diff
@@ -443,8 +472,17 @@ goog.require('ga_urlutils_service');
...
-                gaPreviewFeatures.zoom($scope.map,
-                    geojsonParser.readFeature(f));
+                var feature = geojsonParser.readFeature(f);
+                var ol3d = $scope.ol3d;
+                if (ol3d && ol3d.getEnabled()) {
+                  var extent = feature.getGeometry().getExtent();
+                  extent = gaPreviewFeatures.getMinimalExtent(extent);
+                  ol3d.getCesiumScene().camera.flyTo({
+                    destination: extentToRectangle(extent)
+                  });
```
Comment (oterral) [inline src/components/search/SearchTypesDirectives.js]: Why not putting this in the `gaPreviewFeatures.zoom`  adding ol3d as parameter of the function. We use this function in other place,  I guess.

File: src/components/search/SearchTypesDirectives.js [hunk 3]
```diff
@@ -332,13 +344,29 @@ goog.require('ga_urlutils_service');
...
+                  gaMapUtils.moveTo($scope.map,
+                                    originToZoomLevel[res.attrs.origin],
+                                    [res.attrs.y, res.attrs.x]);
+                }
               } else {
-                gaMapUtils.zoomToExtent($scope.map, e);
+                if (ol3d && ol3d.getEnabled()) {
+                  ol3d.getCesiumScene().camera.flyTo({
+                    destination: extentToRectangle(e)
+                  });
```
Comment (oterral) [inline src/components/search/SearchTypesDirectives.js]: This should be in  `gaMapUtils.zoomToExtent`

File: src/components/search/SearchTypesDirectives.js [hunk 4]
```diff
@@ -332,13 +344,29 @@ goog.require('ga_urlutils_service');
...
-                gaMapUtils.moveTo($scope.map,
-                                  originToZoomLevel[res.attrs.origin],
-                                  [res.attrs.y, res.attrs.x]);
+                if (ol3d && ol3d.getEnabled()) {
+                  ol3d.getCesiumScene().camera.flyTo({
+                    destination: Cesium.Cartesian3.fromDegrees(
+                                   res.attrs.lon,
+                                   res.attrs.lat,
+                                   3000)
+                  });
```
Comment (oterral) [inline src/components/search/SearchTypesDirectives.js]: This should be in `gaMapUtils.moveTo`

**Comment by oterral** [comment]
I've added some comment to avoid duplicate code in other directives. @gjn does it make sense to you to put zoom/move logic of the globe in gaMapUtils functions?

**Comment by gjn** [comment]
+1 for @oterral proposals.

Also, would like to review this today.

**Comment by gjn** [comment]
Fantastic. I like it.

There's an issue with feature search. Add `ch.swisstopo.lubis-luftbilder_farbe` and search for `2010`. Select an object and it flies to this object as expected, but zooms in too much IMO.

Looking at code right now.

File: src/components/MarkerOverlayService.js (hunk 1 above)
Comment (gjn) [inline src/components/MarkerOverlayService.js]: +1

**Comment by gjn** [comment]
Last comment was for 3D only.

For 2D and 3D, the feature preview (when hovering over results in the dropdown menu) changed. It was standard marker before, now it's a blue circle filled with grey. I think it should be standard marker.

**Comment by gjn** [comment]
A couple of things need adaption, otherwise, this is good stuff.

**Comment by oterral** [comment]
jenkins test

**Comment by oterral** [comment]
Ready to review/test

**Comment by cedricmoullet** [comment]
@gjn could you please review and merge ?

**Comment by gjn** [comment]
jenkins test

**Comment by gjn** [comment]
This looks good, but the jenkins failure is indeed valid here. It's a regression. Compare prod to this. Open the link, press the x on the search box. On [prod](https://map.geo.admin.ch/?swisssearch=wasser&lang=de) and in [master](https://mf-geoadmin3.dev.bgdi.ch/?swisssearch=wasser&lang=de), the term disappears from the search box and from the permalink, on [this branch](https://mf-geoadmin3.dev.bgdi.ch/gberaudo_search_layer/?swisssearch=wasser&lang=de) it remains both in the url and in the search box.

**Comment by oterral** [comment]
Tests are good!!  Fixed on the branch

**Comment by gjn** [comment]
Perfect. That's fine work here.

**Pull request merged**